src/pybind/mgr/volumes/fs/health.py

```diff
@@ -0,0 +1,111 @@
+import logging
+import threading
+
+from .exception import VolumeException
+
+log = logging.getLogger(__name__)
+
+# singleton design pattern taken from http://www.aleax.it/5ep.html
+
+class HealthUpdateSingleton(object):
+    """
+    Singleton class to track health metadata for various "modules"
+    (and send update to `ceph status`). Severity of health is left
+    upto the module.
+
+    This is implemented as a singleton to "funnel" health updates
+    for a given type (else, this can be instantiated thousand times
+    and each instance would have a different view of health for a
+    given type resulting in health updates being sent uncoordinated).
+    Yes, this could have been a normal class, instatiated once (at
+    start) and passed around to anyone who wants to send health
+    updates -- but, this is much more general purpose.
+
+    If you subclass this, your class becomes a singleton.
+    """
+
+    VOLUME_HEALTH_PREFIX = "Module 'volumes'"
+
+    # internal identifier(s) for various health attributes
+    PURGE_THREAD_HEALTH = "purge_thread_health"
+
+    # "keys" for various health updates sent to `ceph status`
+    PURGE_THREAD_HEALTH_KEY = "MODULE_VOLUME_PURGE_THREAD"
+
+    # TODO: only send updates for mods that have changed.
+    def _send_health_update(self):
+        health = {}
+        volume_client = self._shared_state['meta']['volume_client']
+
+        for mod_key, func in self.dispatch_table.items():
+            log.info("get health: {0}".format(mod_key))
+            mod, mod_health = func[2](self)
+            health[mod] = mod_health
+        log.debug("sending health={}".format(health))
+        volume_client.send_health_update(health)
+
+    def _purge_thread_health_init(self):
+        self._shared_state['health'][HealthUpdateSingleton.PURGE_THREAD_HEALTH] = []
+
+    def _purge_thread_update_state(self, **kwargs):
+        identifier = kwargs['identifier']
+        remove = kwargs.get('remove', False)
+        if remove:
```

Comment:
when is 'remove=True' ever passed?

Reply:
not as of now since there is no interface in `mgr/volumes` to rectify a failed purge thread.